feature(menus): required AMD modules can now be defined at item registration

hypeJunction opened the pull request:
Adds a "require" parameter to menu item factory, which allows plugins to specify
which AMD modules are to be loaded when the menu item is displayed.

Refs #9514

hypeJunction committed:
feature(menus): required AMD modules can now be defined at item registration

Adds a "require" parameter to menu item factory, which allows plugins to specify
which AMD modules are to be loaded when the menu item is displayed.

Refs #9514

hypeJunction commented on views/default/navigation/menu/elements/required_modules.php:
```diff
@@ -0,0 +1,24 @@
+}
+
+foreach ($modules as $module) {
+	elgg_require_js($module);
```
This won't work with AJAX, but using inline `<script>` would be an overkill, in case of an entity menu, for example.

mrclay replied:
We can teach elgg/Ajax to run these locally

mrclay commented on engine/classes/ElggMenuItem.php:
```diff
@@ -396,6 +399,39 @@ public function addLinkClass($class) {
+	 */
+	public function getRequire() {
+		$modules = (array) $this->data['require'];
+		return array_filter($modules, function($m) {
```
Validate while adding instead.

hypeJunction replied:
Problematic because there are many way in which this can be registered: set
on menu item object property, through factory array or data etc
On Mar 19, 2016 9:44 PM, "Steve Clay" <notifications@github.com> wrote:

> In engine/classes/ElggMenuItem.php
> <https://github.com/Elgg/Elgg/pull/9520#discussion_r56754871>:
>
> > +	 *
> > +	 * @param string[]|string $modules One or more required AMD modules
> > +	 * @return void
> > +	 */
> > +	public function setRequire($modules) {
> > +		$this->data['require'] = (array) $modules;
> > +	}
> > +
> > +	/**
> > +	 * Get required AMD modules
> > +	 *
> > +	 * @return string[]
> > +	 */
> > +	public function getRequire() {
> > +		$modules = (array) $this->data['require'];
> > +		return array_filter($modules, function($m) {
>
> Validate while adding instead.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/Elgg/Elgg/pull/9520/files/31c9e05261f5a0949ddc3b9c6b77d622c092e3e3#r56754871>
>

mrclay replied:
We should validate in __construct() too? Generally we shouldn't invite bad state into an object on assumption we'll clean it later. M

hypeJunction replied:
Maybe even in __set, but need to check if value ends up in data...  or maybe not validate and log errors from elgg_require_js

hypeJunction replied:
I am not going to bother with this, as it will require rewriting the class

mrclay commented:
Rename this feature "requires"

mrclay commented:
Actually better: "deps" like RequireJS uses for shims.

mrclay commented on views/default/navigation/menu/elements/required_modules.php:
```diff
@@ -0,0 +1,24 @@
+	return;
+}
+
+foreach ($modules as $module) {
```
Just `foreach ($item->getDeps() as $module) {`

mrclay commented on engine/lib/navigation.php:
```diff
@@ -613,6 +614,10 @@ function _elgg_nav_init() {
 
 	elgg_register_ajax_view('navigation/menu/user_hover/contents');
+
+	// Using a view extension to ensure that themes that have replaced the item view
```
Smart, and thx for the comment

hypeJunction committed:
perf(reportedcontent): only load JS if menu item is rendered

Reportedcontent would continue to load the JS AMD module even if the extras
menu item was unregistered

Refs #9514

hypeJunction commented:
@mrclay Renamed to `deps`

jdalsem commented:
Can we also include something for other kind of dependencies... as there are multiple menu items that rely on lightbox... Now we need to do: 
```
elgg_load_js('lightbox');
elgg_load_css('lightbox');
``` 
in a register hook....

hypeJunction commented:
I refuse :) Die `elgg_load_js()` a horrible death! PR for converting lightbox to an AMD module is coming eventually. I have all the code that is needed in http://github.com/hypeJunction/elgg_lightbox

jdalsem commented:
if you could make the PR for lightbox as an AMD i would be very glad!

jdalsem commented:
Happy with this PR as is.

mrclay commented on mod/developers/views/default/theme_sandbox/navigation/require.js:
```diff
@@ -0,0 +1,9 @@
+
+	$(document).on('click', '.elgg-menu-require .elgg-menu-item-require > a', function(e) {
+		e.preventDefault();
+		alert('Thank you for trying me');
```
Yay for including a demo!
